### ISSUE TITLE
.gitattributes issue

### ISSUE THREAD

**Issue by davidedelvento** [jonasoreland/runnerup#584]
This is more of a git issue than runnerup issue, however I'm reporting it in case others encounter it. Basically it's a different version of https://stackoverflow.com/questions/5009096/files-showing-as-modified-directly-after-git-clone : when compiling on Linux, `gradlew.bat` is shown as modified and even doing a checkout of the file leaves it in that status. This is minor annoyance in general, but prevents one from using `git stash` effectively (`git stash pop` refuses to execute).
The workaround is to comment out the line `gradlew.bat text eol=crlf` in `.gitattributes` before running `git stash pop`, and then uncommenting it after the fact.

As far as I'm concerned, this may be closed  with `won't fix` status, I'm reporting it just for documenting purposes.

**Comment by gerhardol** [comment]
gradlew.bat is stored as crlf (and gradlew as lf) so the settings in .gitattributes match (there are many .xml stored as crlf but .gitattributes is set to lf though).

How is the file stored after cloning?
What is your setting for core.autocrlf?
(I am on Windows and true/false both works for me).

dos2unix/unix2dos should be a better workaround than editing .gitattributes

If you find what to use, I suggest you make a PR for this.

**Comment by davidedelvento** [comment]
@gerhardol thanks for the message. I don't have that option set, see below.

```
/tmp $ git config --get core.autocrlf
/tmp $ ls runnerup
ls: cannot access runnerup: No such file or directory
/tmp $ git clone git@github.com:jonasoreland/runnerup.git
Cloning into 'runnerup'...
remote: Counting objects: 19117, done.
remote: Total 19117 (delta 0), reused 0 (delta 0), pack-reused 19117
Receiving objects: 100% (19117/19117), 6.42 MiB | 0 bytes/s, done.
Resolving deltas: 100% (10846/10846), done.
Checking connectivity... done.
/tmp $ cd runnerup/
/tmp/runnerup $ git config --get core.autocrlf
/tmp/runnerup $ git status
On branch master
Your branch is up-to-date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   gradlew.bat

no changes added to commit (use "git add" and/or "git commit -a")
$ git diff | wc -l
185
/tmp/runnerup $ git diff -w
/tmp/runnerup $ wc -l gradlew.bat 
90 gradlew.bat
```

Deleting the runnerup directory, running either `git config --global core.autocrlf true` or `false` and repeating the whole process, does not change the outcome by a bit.

I agree, there should be a better way, and there probably is. However life is too short for me to spend time to debug this issue, I prefer to spend my limited time on the other more interesting ones I'm working on.

Cheers

**Comment by gerhardol** [comment]
Yes, we do not want spend time on this, but there should be a better workaround

git --version
How is gradlew.bat stored after clone, with crlf or lf?
git check-attr -a -- gradlew.bat
Is unix2dos (or dos2unix) a workaround?

**Comment by davidedelvento** [comment]
```
/tmp/runnerup $ git --version
git version 1.9.1

/tmp/runnerup $ file gradlew.bat 
gradlew.bat: ASCII text, with CRLF line terminators

/tmp/runnerup $ git check-attr -a -- gradlew.bat
gradlew.bat: text: set
gradlew.bat: eol: crlf
```

> Is unix2dos (or dos2unix) a workaround?

Perhaps, but not in a straightforward way. If I look at the file, as the command said is "dos" format. If I change it to unix, it still displays as modified in `git status` however if now I ignore the whitespace change I see this warning:

```
/tmp/runnerup $ git diff -w
warning: LF will be replaced by CRLF in gradlew.bat.
The file will have its original line endings in your working directory.
```

**Comment by gerhardol** [comment]
/tmp is not handling file mode attributes the same way as normal partitions, that could be a problem
(I have not worked with Linux much in some years, this is a problem in Cygwin at least)

The file modes could be incorrect, I do not know if this could be a problem
https://stackoverflow.com/questions/1580596/how-do-i-make-git-ignore-file-mode-chmod-changes

I suggest this is kept open with the lowest priority for now

**Comment by davidedelvento** [comment]
AFAIK on real linuxes, /tmp is treated equally. It might be treated
differently if it's mounted on a separate partition, as sometimes is (not
on the machines I used for the test).

I've just repeated the very same test in my home directory. I also tried
cloning by https instead of ssh (not because I thought would it make a
difference, but because I did not have the ssh key installed on this other
machine). Very same result, but the version of git was also the same.
Tomorrow I will try from a system with a different version of git.

**Comment by davidedelvento** [comment]
This seems to be an issue with more recent versions of git, in fact `git v1.7.1` works fine:

```
$ git config --get core.autocrlf
$ ls runnerup
ls: cannot access runnerup: No such file or directory
$ git clone git@github.com:jonasoreland/runnerup.git   
Initialized empty Git repository in /glade2/scratch2/ddvento/yellowstone/runnerup/.git/
remote: Counting objects: 19117, done.
remote: Total 19117 (delta 0), reused 0 (delta 0), pack-reused 19117
Receiving objects: 100% (19117/19117), 6.42 MiB | 1.47 MiB/s, done.
Resolving deltas: 100% (10846/10846), done.
$ git --version
git version 1.7.1
$ cd runnerup/
$ git status
# On branch master
nothing to commit (working directory clean)
$ 
```

whereas newer versions, e.g. here `git v2.3.0` or previously `git v1.9.1` (note this is the same machine as above, with the PATH changed to use a different git install)

```
$ git config --get core.autocrlf
$ ls runnerup
ls: cannot access runnerup: No such file or directory
$ git --version
git version 2.3.0
$ git clone git@github.com:jonasoreland/runnerup.git
Cloning into 'runnerup'...
remote: Counting objects: 19117, done.
remote: Total 19117 (delta 0), reused 0 (delta 0), pack-reused 19117
Receiving objects: 100% (19117/19117), 6.42 MiB | 1.43 MiB/s, done.
Resolving deltas: 100% (10846/10846), done.
Checking connectivity... done.
$ cd runnerup
$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

        modified:   gradlew.bat

no changes added to commit (use "git add" and/or "git commit -a")
$ 
```

**Comment by gerhardol** [comment]
Note: master now sets attributes on gradlew.bat
(Updating caused problems for @12people hopefully fixed).
Close?
